pkg/iss/pbftslot.go

```diff
@@ -93,14 +93,23 @@ func (slot *pbftSlot) advanceState(pbft *pbftInstance, sn t.SeqNr) *events.Event
 		// Mark slot as committed.
 		slot.Committed = true
 
-		// Unless everything has already been committed, set a new batch timeout.
+		// Set a new batch timeout.
+		// Note that we set a new batch timeout even if everything has already been committed.
+		// In such a case, we expect a quorum of other nodes to also commit everything and send a Done message
+		// before the timeout fires.
 		// The timeout event contains the current view and the number of committed slots.
-		// It will be ignored if any of those values change by the time the timer fires.
-		if !pbft.allCommitted() {
-			eventsOut.PushBack(pbft.eventService.TimerDelay(
-				t.TimeDuration(pbft.config.ViewChangeBatchTimeout),
-				pbft.eventService.SBEvent(PbftViewChangeBatchTimeout(pbft.view, pbft.numCommitted(pbft.view))),
-			))
+		// It will be ignored if any of those values change by the time the timer fires
+		// or if a quorum of nodes confirms having committed all batches.
+		eventsOut.PushBack(pbft.eventService.TimerDelay(
+			t.TimeDuration(pbft.config.ViewChangeBatchTimeout),
+			pbft.eventService.SBEvent(PbftViewChangeBatchTimeout(pbft.view, pbft.numCommitted(pbft.view))),
+		))
```

Comment:
If we have now the end-of-segment checkpoints then why do we start the batch timeout even though all the slots are committed?

Reply:
Because all the slots might be committed only locally. We need the timeout for the situation where only a few nodes committed all slots locally, but not enough to have a stable segment-level checkpoint. In that case, even those that committed all the slots are necessary for starting a view change protocol that the other correct nodes need. Only when the segment-level checkpoint is stable can we start ignoring timers.

That said, technically, it would not be strictly necessary, since there is still the end-of-segment timeout, but continuing with the batch timer seems rather natural to me.

Reply:
How about the following?

```go
if !pbft.allCommitted() || !pbft.segmentCheckpoint.Stable(len(pbft.segment.Membership)) {
	eventsOut.PushBack(pbft.eventService.TimerDelay(
		t.TimeDuration(pbft.config.ViewChangeBatchTimeout),
		pbft.eventService.SBEvent(PbftViewChangeBatchTimeout(pbft.view, pbft.numCommitted(pbft.view))),
	))
}
```

Reply:
Technically that check is redundant if one looks purely at correctness, since it still must be repeated when the timer fires. Checking here is an optimization to potentially save ourselves a timer event that we already know will be ignored. But it probably also improves the understandability of the code, showing explicitly that we don't need the timeout if the checkpoint is stable.

Bottom line: I added the check :-)
(I also updated the semantics of "stability" of a segment-level checkpoint to be consistent with iss-level checkpoints and require not only having received a quorum of Done messages, but also having committed all slots. That's why I'm now only checking for checkpoint stability.)